examples/game/breakout.rs

```diff
@@ -5,268 +5,381 @@ use bevy::{
     sprite::collide_aabb::{collide, Collision},
 };
 
-/// An implementation of the classic game "Breakout"
-const TIME_STEP: f32 = 1.0 / 60.0;
+use components::*;
+use resources::*;
+
+/// Constants that can be used to fine-tune the behavior of our game
+mod config {
+    use bevy::render::color::Color;
+    use bevy::ui::Val;
+    // TODO: add various Vec2's and Transforms to this config module for clarity and consistency
+    // Blocked on https://github.com/bitshifter/glam-rs/issues/76
+
+    pub const TIME_STEP: f32 = 1.0 / 60.0;
+    pub const BACKGROUND_COLOR: Color = Color::rgb(0.9, 0.9, 0.9);
+
+    pub const PADDLE_COLOR: Color = Color::rgb(0.5, 0.5, 1.0);
+    pub const PADDLE_SPEED: f32 = 500.0;
+
+    pub const BALL_COLOR: Color = Color::rgb(1.0, 0.5, 0.5);
+
+    pub const WALL_THICKNESS: f32 = 10.0;
+    pub const WALL_COLOR: Color = Color::rgb(0.8, 0.8, 0.8);
+
+    pub const BRICK_COLOR: Color = Color::rgb(0.5, 0.5, 1.0);
+
+    pub const SCOREBOARD_COLOR: Color = Color::rgb(0.5, 0.5, 1.0);
+    pub const SCORE_COLOR: Color = Color::rgb(1.0, 0.5, 0.5);
+    pub const SCORE_FONT_SIZE: f32 = 40.0;
+    pub const SCORE_PADDING: Val = Val::Px(5.0);
+}
+
+/// A simple implementation of the classic game "Breakout"
 fn main() {
     App::build()
         .add_plugins(DefaultPlugins)
-        .insert_resource(Scoreboard { score: 0 })
-        .insert_resource(ClearColor(Color::rgb(0.9, 0.9, 0.9)))
-        .add_startup_system(setup.system())
+        .insert_resource(ClearColor(config::BACKGROUND_COLOR))
+        // This adds the Score resource with its default values: 0
+        .init_resource::<Score>()
+        // These systems run only once, before all other systems
+        .add_startup_system(spawn_cameras.system())
+        .add_startup_system(spawn_paddle.system())
+        .add_startup_system(spawn_ball.system())
+        .add_startup_system(spawn_walls.system())
+        .add_startup_system(spawn_bricks.system())
+        .add_startup_system(spawn_scoreboard.system())
+        // These systems run repeatedly, whnever the FixedTimeStep's duration has elapsed
         .add_system_set(
             SystemSet::new()
-                .with_run_criteria(FixedTimestep::step(TIME_STEP as f64))
-                .with_system(paddle_movement_system.system())
-                .with_system(ball_collision_system.system())
-                .with_system(ball_movement_system.system()),
+                .with_run_criteria(FixedTimestep::step(config::TIME_STEP as f64))
+                .with_system(kinematics.system().label("kinematics"))
+                .with_system(ball_collision.system().before("kinematics")),
         )
-        .add_system(scoreboard_system.system())
+        // Ordinary systems run every frame
+        .add_system(bound_paddle.system().label("bound_paddle"))
+        .add_system(paddle_input.system().after("bound_paddle"))
```

Comment:
I think these two should be swapped so input is processed before moving the paddle. Otherwise you get a one frame lag.

Reply:
I tried that and you get stuck against the walls due to your velocity getting reset :( I'll see if I can tweak the logic to fix this; I think the trick is that I need to interleave kinematics between them.

Reply:
Fixed!

Reply:
I was not aware that the order of these statements would cause a one frame lag. But then again I'm a new to bevy. Could add a comment in the code why it's important, maybe with a link to a more in depth resource for anyone curious as to why?

Reply:
That's a good choice :) Let me see what I can do.